### ISSUE TITLE
New firmware image format for QSW-M2116P-2T2S devices

### ISSUE THREAD

**Issue by AlexAltea** [marcan/qsw-tools#2]
I've entered the QSW rabbit hole while trying to debug abnormally high CPU usage and temps on my [QSW-M2116P-2T2S](https://www.qnap.com/en/product/qsw-m2116p-2t2s).

First thing I've noticed is that the [two latest available firmware images](https://www.qnap.com/en/download?model=qsw-m2116p-2t2s&category=firmware) have different formats; none of them plain TAR files, so they cannot be extracted/passed to [`decfile.sh`](https://github.com/marcan/qsw-tools/blob/main/decfile.sh). Both were released on the same day, as part of the same update chain. These are:

## `QSW-M2116P-1.1.0.22052.img`

Identical format to all previous QSW-M2116P-2T2S firmware images. The header looks like:

```
00000000: ded5 d4ed 4d4c 7b98 0200 0000 5c00 0000  ....ML{.....\...
00000010: 54ef 2500 6a61 6775 6172 3263 0000 0000  T.%.jaguar2c....
00000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000030: 0000 0000 6a61 6775 6172 3200 0000 0000  ....jaguar2.....
00000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000050: 0000 0000 0700 0000 0100 0000 0100 0000  ................
00000060: 2000 0000 1000 0000 1b0d 8662 39a1 fb75   ..........b9..u
00000070: f4d1 8d78 0b22 ada1 0000 0000 0000 0000  ...x."..........
00000080: 64a9 2200 54a9 2200 fd37 7a58 5a00 0001  d.".T."..7zXZ...
00000090: 6922 de36 0200 2101 0800 0000 d80f 2313  i".6..!.......#.
000000a0: e287 63ef fe5d 004f 6050 0f80 dd97 f03d  ..c..].O`P.....=
000000b0: 2cf2 e113 5a68 f368 3616 6875 22d1 d9ff  ,...Zh.h6.hu"...
```

And there's what looks like a `.tar.xz` signature at 0x88. Obviously not a `.tar` file like QSW firmware images for other devices.

I haven't managed to extract it yet...

## `QSW-M2116P-2.0.0.22052.img`

Format changes. Now the header is `\x24\x51\x5e\xbe\x08\x0a\x46\x05` followed by:

```
MIME-Version: 1.0
Content-Disposition: attachment; filename="smime.p7m"
Content-Type: application/pkcs7-mime; smime-type=signed-data; name="smime.p7m"
Content-Transfer-Encoding: base64

MIIG2QYJKoZIhvcNAQcCoIIGyjCCBsYCAQExDTALBglghkgBZQMEAgEwVAYJKoZI
hvcNAQcBoEcERVNIQTEoUVNXLU0yMTE2UC0yLjAuMC5pbWcpPSBhNDc5NzY3MmIy
[...]
La12wv/Eak8EB0WJMn1gzK9UJRLrMH/o8G3XuG1mfbKezDrLn9j2GFiuYq8burL+
cTy04pT2+xrJdPNZ6j3Fsu5HbOmS9jwQAytGRMU=

-----BEGIN CERTIFICATE-----
MIIDtzCCAp+gAwIBAgIFANEthgUwDQYJKoZIhvcNAQELBQAwgYAxCzAJBgNVBAYT
AlRXMQ8wDQYDVQQIDAZUYWl3YW4xDzANBgNVBAcMBlRhaXBlaTENMAsGA1UECgwE
[...]
hFiyL2wYXaj7JspixGIMgiEH7rDdQzzTOckfILal4B/IVAL7VDb7XNbHj3d9nkIj
WeexPUNm5tjK2MKnFelheLM4ho2hORLgAhgbnKop9VmQEonOMj54bWZuEA==
-----END CERTIFICATE-----
```

Followed by the data.

---

What's surprising is that on 2023-05-17 you release this repository and [announce it on Reddit](https://www.reddit.com/r/qnap/comments/13jv5nc/qswtools_full_management_cli_on_qnap_switches/), and then, QNAP issues a double update 2 weeks later (the actual build is dated 2023-05-22, only 5 days later) where they change the firmware image format. May be a coincidence, but it's rather suspicious.

I have seen a similar double-update pattern on other QSW switches as well, but the firmware image format remains the same:
- https://www.qnap.com/en/download?model=qsw-m2108-2c&category=firmware
- https://www.qnap.com/en/download?model=qsw-m2108-2s&category=firmware

**Comment by AlexAltea** [comment]
Nevermind, I was a sleep-deprived fool yesterday. The only change is that new images are signed (not encrypted!), and then verified via `/usr/bin/verifySignature.sh`.

Just to be clear, both QSW-M2116P-2T2S firmware files are plain redboot jaguar2 images, which can be recursively extracted via `binwalk -e QSW-M2116P-1.1.0.22052.img` and `binwalk -e QSW-M2116P-2.0.0.22052.img` respectively.

I'll have a look at them now, since `guest:guest123` CLI backdoor doesn't seem to work on my QSW-M2116P-2T2S.

**Comment by AlexAltea** [comment]
There's no `guest` account, but the regular admin user gets a privileged ICLI shell.

There's no `debug` account either, so after enabling SSH, you are stuck with the lame ICLI.

I can't seem to get a root /bin/sh shell, or enable debug mode, despite there being some strings in `/usr/bin/switch_app` implying the command should be "_platform debug { allow | deny }_". I cannot modify the platform configuration which is quite frustrating. Any hints would be very much appreciated.

**Comment by sbingner** [comment]
I got a QSW-M3216 and I can log into the shell but can do practically nothing in it.  Also it allowed me to enable ssh but then I couldn't even login over ssh.

Have you had any luck finding a way to get into new firmwares? Tempting to return this thing, the interface is pretty bad and it's really frustrating to see how close something useful is.

EDIT: Actually I broke into this device and it looks like the firmware is completely different.  It's running openwrt and has no switch_app or qnss-web but rather runs `ISS.exe` and `aricent-backend.arm64`  - I made a keygen for the output of CTRL-T to get the password for CTRL-R which is so silly.  The enable mode on this thing isn't great.  I need to clean it up before posting anywhere but reach out if you want it first.

**Comment by AlexAltea** [comment]
Nope, unfortunately I haven't got time to reverse engineer the code related to platform debug.

But I've had a quick glimpse at other firmware images and it seems QNAP has *wildly* different stacks in different switches.

So what you found is likely not applicable to my model, or @marcan's model.

**Comment by psotres** [comment]
Sorry for jumping in.

I'm trying to enable SSH on a QNAP 5216, which seems to follow a similar approach as the one @sbingner mentioned. I've been examining the firmware without much luck. @sbingner could you point me into the right direction on how to generate a new keygen for this model?

**Comment by sbingner** [comment]
> Sorry for jumping in.
> 
> I'm trying to enable SSH on a QNAP 5216, which seems to follow a similar approach as the one @sbingner mentioned. I've been examining the firmware without much luck. @sbingner could you point me into the right direction on how to generate a new keygen for this model?

you can see what I did for the M3216 at https://github.com/sbingner/qsw-tools-arm64/blob/main/genpwd.py

**Comment by junekiwi** [comment]
> I can't seem to get a root /bin/sh shell, or enable debug mode, despite there being some strings in /usr/bin/switch_app implying the command should be "platform debug { allow | deny }". I cannot modify the platform configuration which is quite frustrating. Any hints would be very much appreciated.


`platform debug allow` is not available over SSH but it is over console. See https://github.com/maxmouchet/QSW-M2116P-2T2S?tab=readme-ov-file#shell-access.

**Comment by Civil** [comment]
@sbingner I wonder how you've got the password generation method. I've recently got QSW-M7308R-4X, which runs their QSW 4.x (I've tried their stock 1.4.1, newer 4.0.1 and newest 4.1.1), and while I can get the challenge, it doesn't accept the generated response (neither with -4 nor with default salt)

UPD: I've actually got a root shell via other ways, so I'm trying to figure out where to get a new salt for the password
UPD2: That was faulty USB-RS232 adapter that I had that was adding garbage to the input. Your patched script worked with -4.

**Comment by mikegleasonjr** [comment]
Hey guys, I don't mean to hichjack the thread, but when loggin on my switch (different model), I see 100% cpu usage by the `ISS408wt.exe`, can you check if yours is doing the same? See #5, thanks!
